Vega 15 (#26)
NLESC-JCER/cpp2wasm: TESTING.md

To make sure [JavaScript and WebAssembly code snippets](README.md#JavaScript) and [Single page application](README.md#single-page-application) work we want have a tests for them.

To test, we will use the [cypress](https://www.cypress.io/) JavaScript end to end testing framework.
Cypress can simulate user behavior such as clicking buttons etc. and checks expected result in a web browser.

In the following examples, we test if the example web pages render the answer `-1.00` when they are visited.

> To visit a web page we need to start a web server with

Reviewer: ```suggestion
To visit a web page we need to start a simple web server with using Python
```

```

And lastly, a test for the React/form/Web worker/WebAssembly combination.
Let us also change the guess value.

```{.js file=cypress/integration/example-app_spec.js}
describe('src/js/example-app.html', () => {
  it('should render -1.00', () => {
    cy.visit('http://localhost:8000/src/js/example-app.html');
    cy.get('input[name=guess]').type('0');

>     // TODO assert value is set

Reviewer: Can you make this comment more explicit?

Author: Added comment that explains what type does

});
```

Third, a test for the [React/form/Web worker/WebAssembly combination](http://localhost:8000/src/js/example-app.html).
Let us also change the initial guess value.

```{.js file=cypress/integration/example-app_spec.js}
describe('src/js/example-app.html', () => {
  it('should render -1.00', () => {
    cy.visit('http://localhost:8000/src/js/example-app.html');

>     // In initial guess input field type 0 to append 0 to default -20 so the value is -200

Reviewer: ```suggestion
    // We append a 0 to the guess input field so it becomes -200 
```

    cy.get('input[name=guess]').type('0');
    cy.contains('Submit').click();
    cy.get('#answer').contains('-1.00');
  });
});
```

> And another test same as before, but now with [JSON schema powered form](http://localhost:8000/src/js/example-jsonschema-form.html).

Reviewer: ```suggestion
And similar test to the previous one, but now with [JSON schema powered form](http://localhost:8000/src/js/example-jsonschema-form.html).
```

> And another test for the full application, but now with [JSON schema powered form](http://localhost:8000/src/js/example-jsonschema-form.html).

Author: ```suggestion
And another test same as before, but now with [JSON schema powered form](http://localhost:8000/src/js/example-jsonschema-form.html).
```

```{.js file=cypress/integration/example-jsonschema-form_spec.js}
describe('src/js/example-jsonschema-form.html', () => {
  it('should render -1.00', () => {
    cy.visit('http://localhost:8000/src/js/example-jsonschema-form.html');

>     // The JSON schema powered form uses a hierarchy of identifers for each input field starting with `root`, as the `epsilon` input field is a direct child of root it has `root_epsilon` as an identifier

Reviewer: ```suggestion
    // The JSON schema powered form uses a hierarchy of identifers for each input field starting with `root`
    // As the `epsilon` input field is a direct child of root, it has `root_epsilon` as an identifier
```

});
```

Third, a test for the [React/form/Web worker/WebAssembly combination](http://localhost:8000/src/js/example-app.html).
Let us also change the initial guess value.

```{.js file=cypress/integration/example-app_spec.js}
describe('src/js/example-app.html', () => {
  it('should render -1.00', () => {
    cy.visit('http://localhost:8000/src/js/example-app.html');

>     // We append a 0 to the guess input field so it becomes -200 

Reviewer: ```suggestion
    // The initial value of the guess input field is -20 so we append a 0 and it becomes -200 
```

    cy.get('input[name=guess]').type('0');
    cy.contains('Submit').click();
    cy.get('#answer').contains('-1.00');
  });
});
```

And similar test to the previous one, but now with [JSON schema powered form](http://localhost:8000/src/js/example-jsonschema-form.html).

```{.js file=cypress/integration/example-jsonschema-form_spec.js}
describe('src/js/example-jsonschema-form.html', () => {
  it('should render -1.00', () => {
    cy.visit('http://localhost:8000/src/js/example-jsonschema-form.html');
    // The JSON schema powered form uses a hierarchy of identifiers for each input field starting with `root`
    // As the `epsilon` input field is a direct child of root, it has `root_epsilon` as an identifier
    const input_selector = 'input[id=root_epsilon]';

>     // In initial guess input field replace default value of initial guess with 0.1

Reviewer: ```suggestion
    // In initial guess input field we replace the default value with 0.1
```
